### PR TITLE
GROOVY-8360: Enums that are inner classes should have the static clas…

### PR DESCRIPTION
**Pull request by shils**
…s modifier

### CODE REVIEW

**Comment by paulk-asert** [comment]
On the one hand, the JLS[1] says: "_A nested enum type is implicitly_ `static`", so your proposed change looks good. However it goes on to say that "_This implies that it is impossible to declare an enum type in the body of an inner class_". With some simple experimenting, it seems Groovy doesn't have this same restriction. I think we just need to do a little more checking that such cases aren't adversely affected (or understand if they are). I'll try to play around a little more myself but have other things on the TODO list first.

[1] [8.9. Enum Types](http://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.9)

**Comment by shils** [comment]
@paulk-asert That's an interesting case. I added a few variations of my original test to cover that along with static compilation. Looking at the byte code, enums nested within inner classes don't seem to have any references to instances of their enclosing classes, so I'd *think* this would be ok. I'll try to test this with some ecosystem projects when I have time.

File: src/test/gls/enums/EnumTest.groovy [hunk 1]
```diff
@@ -592,6 +592,80 @@ class EnumTest extends CompilableTestSupport {
...
+                enum Bar {
+                    X('x'), Y
+                    String s
+                    Bar(String s) { this.s = s }
+                    Bar() {}
+                }
+            }          
+            @groovy.transform.CompileStatic
+            void test() {
+                assert java.lang.reflect.Modifier.isStatic(Foo.Bar.getModifiers())
```
Comment (shils) [inline src/test/gls/enums/EnumTest.groovy]: Unrelated to this issue - property style getter calls on Enum classes throw NoSuchFieldErrors at runtime when statically compiled.

Comment (paulk-asert) [reply src/test/gls/enums/EnumTest.groovy]: I created GROOVY-8369 for this. I can't at first glance see a reason why we'd not support this with @CS.

**Comment by paulk-asert** [comment]
+1 for this to be merged to at least 2_5_X. I suspect there are further cases which differ from Java which need to be clarified but I think this takes us in the right direction and it will be further locking down to rule out cases that don't really make sense that Groovy appears to support but doesn't really and can be done outside the scope of this PR.

**Pull request merged**